### ISSUE TITLE
Invalid SPIR-V generation with types differing in ArrayStride decoration

### ISSUE THREAD

**Issue by ShabbyX** [KhronosGroup/glslang#2637]
Note: Copied from KhronosGroup/SPIRV-Tools#4273

We encountered a SPIR-V validation error in ANGLE, and it's seems to be due to a glslang bug. Please see issuetracker.google.com/182568639 where you can find the SPIR-V that's failing validation (in comment 8) + my analysis (in comment 9).

The short of it is the following. The source shader has a uniform block containing a vec3[9]. This is given type %168 with decoration ArrayStride 16. There's a function that expects a vec3[9], which is given type %93. The function call site does:

%941 = OpLoad %168 ...
... = OpFunctionCall ... %941 ...
SPIR-V validation is complaining that:

OpFunctionCall Argument <id> '941[%941]'s type does not match Function <id> '93[%_arr_v3float_uint_9]'s parameter type
In other words, it's passed a parameter of type %168, but expected type %93. glslang should have copied the object to the correct type.

**Comment by greg-lunarg** [comment]
It seems to me the solution is to do an OpCopyLogical of %941 with a resulting type %93 and pass the result of that to the function call.

I will implement the above solution as soon as possible, but if someone has already begun a solution, please leave a message here to avoid duplicate effort.

**Comment by greg-lunarg** [comment]
Cleaned up template for GLSL (from google issue tracker:

```
uniform highp vec3 u[9];

highp vec3 f(const highp vec3 u[9], ...){
    ... uses u ...
}

... f(u, ...));
```

**Comment by ShabbyX** [comment]
@greg-lunarg I'm not sure if that's really an issue. AFAIK, drivers end up inlining everything to maximize their ability to reorder instructions, which means there is no actual copy involved. Even then the copies like this are effectively "register renaming" (because the original copy is unused afterwards).

That said, this is ANGLE, and the shaders come from the application, so "not my fault" :D

Whether there's another way or not is beyond me. If you have this readily available, you could make glslang inline such a function before you generate SPIR-V and avoid the problem in the first place.

**Comment by greg-lunarg** [comment]
OK. I didn't know if you had control over that shader or not. Thought I would at least try :)

**Comment by ShabbyX** [comment]
This bug is probably not entirely fixed. I can see this bug manifesting itself in the following situations:

1. An array member of an interface block passed to a function (which is fixed by your change)
2. A struct member of an interface block passed to a function (note that structs also generate multiple types based on std140, std430 or local)
3. A struct member of an interface block assigned to a local struct
4. A struct member of an interface block assigned to a struct member of another interface block with different storage (std140 vs std430)
5. Same as 1 and 2, but used in a constructor (rather than function call)

**Comment by ShabbyX** [comment]
Note also that you could have arrays of structs, or structs containing arrays etc, which the copy code should be able to handle.

**Comment by greg-lunarg** [comment]
The code I enabled recursively process data structures, so recursive objects should work.

I will try to double back on this when I have time, but if you have shaders that you think are important and are having issues, please open an issue. Thanks.
